src/libcore/marker.rs

```diff
@@ -672,6 +673,49 @@ mod impls {
     unsafe impl<T: Send + ?Sized> Send for &mut T {}
 }
 
+/// Compiler-internal trait used to indicate the type of enum discriminants.
+///
+/// This trait is automatically implemented for every type and does not add any
+/// guarantees to [`mem::Discriminant`]. It is **undefined behavior** to transmute
+/// between `DiscriminantKind::Discriminant` and `mem::Discriminant`.
+///
+/// [`mem::Discriminant`]: https://doc.rust-lang.org/stable/core/mem/struct.Discriminant.html
+#[unstable(
+    feature = "discriminant_kind",
+    issue = "none",
+    reason = "this trait is unlikely to ever be stabilized, use `mem::discriminant` instead"
+)]
+#[cfg_attr(not(bootstrap), lang = "discriminant_kind")]
+pub trait DiscriminantKind {
+    /// The type of the dicriminant, which must satisfy the trait
+    /// bounds required by `mem::Discriminant`.
+    type Discriminant: Clone + Copy + Debug + Eq + PartialEq + Hash + Send + Sync + Unpin;
+}
+
+// This default impl is always overwritten by the compiler.
+//
+// We have to use specialization here, as we want to allow
+// `<T as DiscriminantKind>::Discriminant` without adding a trait bound
+// to `T`. We can't just add a builtin implementation for each type on the
+// fly, as the concrete associated type is not always known during typeck.
```

Comment:
What if we assume any type implements this trait, but then *refuse to project* if it's not known enough?
So so `T: DiscriminantKind` would always succeed but `<T as DiscriminantKind>::Discriminant` wouldn't always normalize.

cc @nikomatsakis would this break anything?

Reply:
Don't know how this should be implemented. I also don't think that it would be a lot cleaner than the current approach.

Reply:
> Don't know how this should be implemented.

Okay I just looked and: make `select.rs` always succeed but in `project.rs` change your match arm for `let eligible = match &vtable {` to produce `false` when the type isn't known enough.

This is both cleaner and matches what we do for specialization.